core/tls-websocket-server/src/ws_server.rs

```diff
@@ -0,0 +1,86 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use crate::{
+	common::make_config, connection::TungsteniteWsConnection, WebSocketError, WebSocketHandler,
+	WebSocketResult, WebSocketServer,
+};
+use log::*;
+use std::{net, net::TcpListener, string::String, sync::Arc};
+
+/// Secure web-socket server implementation using the tungstenite library
+pub struct TungsteniteWsServer {
+	ws_address: String,
+	cert_path: String,
+	private_key_path: String,
+}
+
+impl TungsteniteWsServer {
+	pub fn new(ws_address: String, cert_path: String, private_key_path: String) -> Self {
+		TungsteniteWsServer { ws_address, cert_path, private_key_path }
+	}
+}
+
+impl WebSocketServer for TungsteniteWsServer {
+	type Connection = TungsteniteWsConnection;
+
+	fn run<Handler>(&self, handler: Arc<Handler>) -> WebSocketResult<()>
+	where
+		Handler: WebSocketHandler<Connection = Self::Connection>,
+	{
+		debug!("Running tungstenite web socket server on {}", self.ws_address);
+
+		let socket_addr: net::SocketAddr =
+			self.ws_address.parse().map_err(WebSocketError::InvalidWsAddress)?;
+
+		let config = make_config(self.cert_path.as_str(), self.private_key_path.as_str())?;
+
+		let listener = TcpListener::bind(&socket_addr).map_err(WebSocketError::TcpBindError)?;
+
+		loop {
+			let stream_result = listener.accept();
+
+			match stream_result {
+				Ok((stream, _)) => {
+					let cloned_config = config.clone();
+
+					let server_session = rustls::ServerSession::new(&cloned_config);
+
+					let connection = match TungsteniteWsConnection::connect(stream, server_session)
+					{
+						Ok(c) => c,
+						Err(e) => {
+							error!("failed to establish web-socket connection: {:?}", e);
+							continue
+						},
+					};
+
+					// continue serving requests, even if there is an error in handling a specific connection
+					if let Err(handler_error) = handler.handle(connection) {
+						error!("web-socket request failed: {:?}", handler_error);
+					}
+				},
+				Err(e) => {
+					warn!("failed to establish web-socket connection ({:?})", e)
+				},
+			}
+		}
```

Comment:
We loop to accept connections from clients. This is done in a single thread so far. This means no concurrent connections (as was the case in the previous implementation too). If the server is busy serving one client, no new connections are accepted until the current session is closed.

What we'd probably want to have in the future, is a thread pool, onto which we can schedule connections until all threads from the thread pool are exhausted. The size of the thread pool is limited by the number of physical cores (SGX limitation).

Reply:
That is my understanding, yes. The trusted operation pool is processed on a different thread, and `handle()` here returns as soon as the first immediate call is processed and passes the connection into the TOP (if necessary).